### PR TITLE
[SPARK-16334] Maintain single dictionary per row-batch in vectorized parquet reader

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

As part of the bugfix in https://github.com/apache/spark/pull/12279, if a row batch consist of both dictionary encoded and non-dictionary encoded pages, we explicitly decode the dictionary for the values that are already dictionary encoded. Currently we reset the dictionary while reading every page that can potentially cause `java.lang.ArrayIndexOutOfBoundsException` while decoding older pages. This patch fixes the problem by maintaining a single dictionary per row-batch in vectorized parquet reader.
## How was this patch tested?

Manual Tests against a number of hand-generated parquet files.

### CODE REVIEW

**Comment by sameeragarwal** [comment]
cc @ericl

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 1]
```diff
@@ -146,10 +153,8 @@ void readBatch(int total, ColumnVector column) throws IOException {
       int num = Math.min(total, leftInPage);
       if (useDictionary) {
         // Read and decode dictionary ids.
-        ColumnVector dictionaryIds = column.reserveDictionaryIds(total);
         defColumn.readIntegers(
             num, dictionaryIds, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
-
         if (column.hasDictionary() || (rowId == 0 &&
```
Comment (ericl) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: In the original fix, do you know the reasoning behind calling `column.setDictionary()` when the column already has an existing dictionary? That seems incorrect to me.

Comment (sameeragarwal) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: yes, that didn't seem right to me. But it'd still be a no-op.

**Comment by ericl** [comment]
Looks good to me.

tldr: this PR fixes the [non-encoded, encoded, non-encoded] case, but there is still another issue if the row batch ends with an encoded page, e.g. [non-encoded, encoded, encoded]. Though we were not able to reproduce such a parquet file displaying this behavior.

**Comment by sameeragarwal** [comment]
Correct, and here is an explicit test for a separate [encoded, non-encoded] case: https://github.com/apache/spark/pull/14304

**Comment by rxin** [comment]
Merging in master/2.0.
